Error while checking build/O2Physics/o2 for 64fbe7083db399e02bb4761b7831908c136bf08c at 2026-08-16 13:06:
```
## sw/BUILD/lhapdf-latest/log
collect2: error: ld returned 1 exit status


## sw/BUILD/O2Physics-latest/log
2026-08-16T13:03:10.971+0200 [37837:140176140560960] [buildboxcommon_grpcretrier.cpp:177] [ERROR] Retry limit (0) exceeded for "ByteStream.Read()", last gRPC error was [14: recvmsg:Connection reset by peer]
2026-08-16T13:03:11.044+0200 [37837:140176140560960] [executioncontext.cpp:739] [ERROR] Failed to download output blob 5cacbae188cf1e0c00541e49f78288c78e0a62d61aa2e5cebf65b53cf8b1ba95: 14: recvmsg:Connection reset by peer
2026-08-16T13:03:10.972+0200 [33778:140390272567872] [buildboxcommon_grpcretrier.cpp:177] [ERROR] Retry limit (0) exceeded for "FindMissingBlobs()", last gRPC error was [14: recvmsg:Connection reset by peer]
2026-08-16T13:03:10.971+0200 [37233:139799818869312] [buildboxcommon_grpcretrier.cpp:177] [ERROR] Retry limit (0) exceeded for "ByteStream.Read()", last gRPC error was [14: recvmsg:Connection reset by peer]
2026-08-16T13:03:11.044+0200 [37233:139799818869312] [executioncontext.cpp:739] [ERROR] Failed to download output blob 22de65109d1535944a862a01b981b1b380126f1b7b51abbcad10e99786a40ee9: 14: recvmsg:Connection reset by peer
2026-08-16T13:03:10.971+0200 [37642:139721388722752] [buildboxcommon_grpcretrier.cpp:177] [ERROR] Retry limit (0) exceeded for "ByteStream.Read()", last gRPC error was [14: Socket closed]
2026-08-16T13:03:11.044+0200 [37642:139721388722752] [executioncontext.cpp:739] [ERROR] Failed to download output blob 1e3845a3eff4c894b9cbed26895b10205e37a8c03c8a4b48ab4a092a67065b22: 14: Socket closed
2026-08-16T13:03:10.972+0200 [37811:139685008915008] [buildboxcommon_grpcretrier.cpp:177] [ERROR] Retry limit (0) exceeded for "ByteStream.Read()", last gRPC error was [14: Socket closed]
2026-08-16T13:03:11.044+0200 [37811:139685008915008] [executioncontext.cpp:739] [ERROR] Failed to download output blob c868766807c6141a5e1331cdf9b2ea50d177f9aeeaa0e6990ae155ac1758c468: 14: Socket closed
2026-08-16T13:03:10.972+0200 [37887:140517956381248] [buildboxcommon_grpcretrier.cpp:177] [ERROR] Retry limit (0) exceeded for "ByteStream.Read()", last gRPC error was [14: Socket closed]
2026-08-16T13:03:11.044+0200 [37887:140517956381248] [executioncontext.cpp:739] [ERROR] Failed to download output blob 3ca1f08272c717b2f34dc4269181a3c037a2f160f94cd5fae29aaff6a076a58b: 14: Socket closed
2026-08-16T13:03:10.971+0200 [37946:139767531543104] [buildboxcommon_grpcretrier.cpp:177] [ERROR] Retry limit (0) exceeded for "ByteStream.Read()", last gRPC error was [14: Socket closed]
2026-08-16T13:03:11.044+0200 [37946:139767531543104] [executioncontext.cpp:739] [ERROR] Failed to download output blob 7ca5d3ff4b67c4c0159a7067b3141b36a50dc772624a4114baed1c084c8a1832: 14: Socket closed
2026-08-16T13:03:10.972+0200 [37972:139767895853632] [buildboxcommon_grpcretrier.cpp:177] [ERROR] Retry limit (0) exceeded for "ByteStream.Read()", last gRPC error was [14: Socket closed]
2026-08-16T13:03:11.044+0200 [37972:139767895853632] [executioncontext.cpp:739] [ERROR] Failed to download output blob 2a567232460ac1cc1d432ae6d1f03ab7077965029a812d21ea584cdae4f8ac51: 14: Socket closed
2026-08-16T13:03:10.971+0200 [38075:140571864140352] [buildboxcommon_grpcretrier.cpp:177] [ERROR] Retry limit (0) exceeded for "ByteStream.Read()", last gRPC error was [14: Socket closed]
2026-08-16T13:03:11.044+0200 [38075:140571864140352] [executioncontext.cpp:739] [ERROR] Failed to download output blob bbdeee7c4a6c56080cac85cc3e146eb2439df86dd938071344b414e562dbfabe: 14: Socket closed
2026-08-16T13:03:10.972+0200 [38100:139634357713472] [buildboxcommon_grpcretrier.cpp:177] [ERROR] Retry limit (0) exceeded for "ByteStream.Read()", last gRPC error was [14: Socket closed]
2026-08-16T13:03:11.044+0200 [38100:139634357713472] [executioncontext.cpp:739] [ERROR] Failed to download output blob 5ffee1f99684a78ff043a52337a81983e6bc1f227c4be0a0104a3517d529d509: 14: Socket closed
2026-08-16T13:03:10.971+0200 [38127:140072093407808] [buildboxcommon_grpcretrier.cpp:177] [ERROR] Retry limit (0) exceeded for "ByteStream.Read()", last gRPC error was [14: Socket closed]
2026-08-16T13:03:11.044+0200 [38127:140072093407808] [executioncontext.cpp:739] [ERROR] Failed to download output blob 20fe2b98c9b632552208cd1772233f76f344273b39dd23586d3e8c741fd4e3e0: 14: Socket closed
2026-08-16T13:03:10.972+0200 [38129:140671801932352] [buildboxcommon_grpcretrier.cpp:177] [ERROR] Retry limit (0) exceeded for "ByteStream.Read()", last gRPC error was [14: Socket closed]
2026-08-16T13:03:11.044+0200 [38129:140671801932352] [executioncontext.cpp:739] [ERROR] Failed to download output blob 2b868dcf046c677df323cf8b04604145d50d09b63af38b9e52a2df297020c0b1: 14: Socket closed
2026-08-16T13:03:10.971+0200 [38132:140526447819328] [buildboxcommon_grpcretrier.cpp:177] [ERROR] Retry limit (0) exceeded for "ByteStream.Read()", last gRPC error was [14: Socket closed]
2026-08-16T13:03:11.044+0200 [38132:140526447819328] [executioncontext.cpp:739] [ERROR] Failed to download output blob d2eebaf3ff933ab90333e75116c91f232ea1087f489d37ac56d6b2b524c2c4f3: 14: Socket closed
2026-08-16T13:03:10.972+0200 [38198:140241849846336] [buildboxcommon_grpcretrier.cpp:177] [ERROR] Retry limit (0) exceeded for "ByteStream.Read()", last gRPC error was [14: Socket closed]
2026-08-16T13:03:11.044+0200 [38198:140241849846336] [executioncontext.cpp:739] [ERROR] Failed to download output blob f6846148859947d9705b7c860de0cc15fef48f7ef97f98c0fab90eae7c0df7f6: 14: Socket closed
2026-08-16T13:03:10.971+0200 [38221:139876876982848] [buildboxcommon_grpcretrier.cpp:177] [ERROR] Retry limit (0) exceeded for "ByteStream.Read()", last gRPC error was [14: Socket closed]
2026-08-16T13:03:11.044+0200 [38221:139876876982848] [executioncontext.cpp:739] [ERROR] Failed to download output blob 8f1ee55e2f353a8cbaaf1e0440a65a7a6a3f6764c10de9884047b76b652a248e: 14: Socket closed
2026-08-16T13:03:10.972+0200 [38248:140481673189952] [buildboxcommon_grpcretrier.cpp:177] [ERROR] Retry limit (0) exceeded for "ByteStream.Read()", last gRPC error was [14: Socket closed]
2026-08-16T13:03:11.044+0200 [38248:140481673189952] [executioncontext.cpp:739] [ERROR] Failed to download output blob 6848884c751746b5c0a10ee2270a83979bf8f8c6b6bc2b7e7d9b029d9e0564e0: 14: Socket closed
2026-08-16T13:03:10.972+0200 [38276:139768357886528] [buildboxcommon_grpcretrier.cpp:177] [ERROR] Retry limit (0) exceeded for "ByteStream.Read()", last gRPC error was [14: Socket closed]
2026-08-16T13:03:11.044+0200 [38276:139768357886528] [executioncontext.cpp:739] [ERROR] Failed to download output blob 178ab78f4c3ac5da26a40a9efdaabbdc1be7bf8905bc7149210a74e732b63986: 14: Socket closed
2026-08-16T13:03:10.972+0200 [38335:140235121605184] [buildboxcommon_grpcretrier.cpp:177] [ERROR] Retry limit (0) exceeded for "ByteStream.Read()", last gRPC error was [14: Socket closed]
2026-08-16T13:03:11.044+0200 [38335:140235121605184] [executioncontext.cpp:739] [ERROR] Failed to download output blob 3179e21da740623a02d163841400a8289acb81c73af8da5f56420e24d3e3fe1f: 14: Socket closed
2026-08-16T13:03:10.972+0200 [38357:139645610400320] [buildboxcommon_grpcretrier.cpp:177] [ERROR] Retry limit (0) exceeded for "BatchReadBlobs()", last gRPC error was [14: Socket closed]
2026-08-16T13:03:10.988+0200 [38357:139645610400320] [buildboxcommon_grpcretrier.cpp:177] [ERROR] Retry limit (0) exceeded for "ByteStream.Read()", last gRPC error was [14: failed to connect to all addresses; last error: UNKNOWN: ipv4:127.0.0.1:8980: Failed to connect to remote host: Connection refused]
2026-08-16T13:03:11.044+0200 [38357:139645610400320] [executioncontext.cpp:739] [ERROR] Failed to download output blob e8073896b53ef703c05ce52c99a441045c784ea98463d7b44d4536e0f89bf1c2: 14: failed to connect to all addresses; last error: UNKNOWN: ipv4:127.0.0.1:8980: Failed to connect to remote host: Connection refused
2026-08-16T13:03:10.972+0200 [38418:140706569751104] [buildboxcommon_grpcretrier.cpp:177] [ERROR] Retry limit (0) exceeded for "ActionCache.GetActionResult()", last gRPC error was [14: Socket closed]
2026-08-16T13:03:11.044+0200 [38418:140706569751104] [executioncontext.cpp:544] [ERROR] Error while querying action cache at "http://localhost:8980": 14: Socket closed
2026-08-16T13:03:10.972+0200 [38469:139662797183552] [buildboxcommon_grpcretrier.cpp:177] [ERROR] Retry limit (0) exceeded for "ActionCache.GetActionResult()", last gRPC error was [14: Socket closed]
2026-08-16T13:03:11.044+0200 [38469:139662797183552] [executioncontext.cpp:544] [ERROR] Error while querying action cache at "http://localhost:8980": 14: Socket closed
2026-08-16T13:03:10.972+0200 [38491:140596414933568] [buildboxcommon_grpcretrier.cpp:177] [ERROR] Retry limit (0) exceeded for "ActionCache.GetActionResult()", last gRPC error was [14: Socket closed]
2026-08-16T13:03:11.044+0200 [38491:140596414933568] [executioncontext.cpp:544] [ERROR] Error while querying action cache at "http://localhost:8980": 14: Socket closed
[0 more errors; see full log]
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/O2Physics/17431/64fbe7083db399e02bb4761b7831908c136bf08c/build_O2Physics_o2/pretty.html).

--- a/PWGCF/Femto/Core/cascadeBuilder.h
+++ b/PWGCF/Femto/Core/cascadeBuilder.h
@@ -65,7 +65,8 @@ struct ConfCascadeFilters : o2::framework::ConfigurableGroup {
 // NOLINTNEXTLINE(cppcoreguidelines-macro-usage)
 #define CASCADE_DEFAULT_BITS                                                                                                                                           \
   o2::framework::Configurable<bool> passThrough{"passThrough", false, "If true, all Cascades are passed through. Bits for all selections are stored."};                \
-  o2::framework::Configurable<std::vector<float>> cascadeCpaMin{"cascadeCpaMin", {0.95f}, "Minimum cosine of pointing angle"};                                         \
+  o2::framework::Configurable<std::vector<std::string>> cascadeCpaMin{"cascadeCpaMin", {"0.95"}, "Minimum CPA as TFormula, x=pt"};                                     \
+  o2::framework::Configurable<std::vector<std::string>> cascadePaMax{"cascadePaMax", {}, "Maximum pointing angle (rad) as TFormula, x=pt. Empty to disable"};          \
   o2::framework::Configurable<std::vector<float>> cascadeTransRadMin{"cascadeTransRadMin", {0.9f}, "Minimum transverse radius (cm)"};                                  \
   o2::framework::Configurable<std::vector<float>> cascadeDcaDauMax{"cascadeDcaDauMax", {0.25f}, "Maximum DCA between the daughters at decay vertex (cm)"};             \
   o2::framework::Configurable<std::vector<float>> lambdaCpaMin{"lambdaCpaMin", {0.78f}, "Minimum cosine of pointing angle"};                                           \
@@ -128,6 +129,7 @@ struct ConfOmegaSelection : o2::framework::ConfigurableGroup {
 enum CascadeSels {
   // selections for cascades
   kCascadeCpaMin,      ///< Min. CPA (cosine pointing angle)
+  kCascadePaMax,       ///< Max. PA (pointing angle in rad); disabled by default
   kCascadeDcaDaughMax, ///< Max. DCA of the daughers at decay vertex
   kCascadeTransRadMin, ///< max. transverse radius
 
@@ -162,6 +164,7 @@ constexpr char OmegaSelHistName[] = "hOmegaSelection";
 constexpr char CascadeSelsName[] = "Cascade Selection Object";
 const std::unordered_map<CascadeSels, std::string> cascadeSelectionNames = {
   {kCascadeCpaMin, "Cascade CPA Min"},
+  {kCascadePaMax, "Cascade PA Max"},
   {kCascadeDcaDaughMax, "Cascade DCA Daughters Max"},
   {kCascadeTransRadMin, "Cascade Transverse Radius Min"},
 
@@ -270,7 +273,8 @@ class CascadeSelection : public baseselection::BaseSelection<float, o2::analysis
     this->addSelection(kPosDauTof, cascadeSelectionNames.at(kPosDauTof), config.posDauTof.value, limits::kAbsUpperLimit, true, mRequireTof, false);
     this->addSelection(kNegDauTof, cascadeSelectionNames.at(kNegDauTof), config.negDauTof.value, limits::kAbsUpperLimit, true, mRequireTof, false);
 
-    this->addSelection(kCascadeCpaMin, cascadeSelectionNames.at(kCascadeCpaMin), config.cascadeCpaMin.value, limits::kLowerLimit, true, true, false);
+    this->addSelection(kCascadeCpaMin, cascadeSelectionNames.at(kCascadeCpaMin), filter.ptMin.value, filter.ptMax.value, config.cascadeCpaMin.value, limits::kLowerFunctionLimit, true, true, false);
+    this->addSelection(kCascadePaMax, cascadeSelectionNames.at(kCascadePaMax), filter.ptMin.value, filter.ptMax.value, config.cascadePaMax.value, limits::kUpperFunctionLimit, true, true, false);
     this->addSelection(kCascadeTransRadMin, cascadeSelectionNames.at(kCascadeTransRadMin), config.cascadeTransRadMin.value, limits::kLowerLimit, true, true, false);
     this->addSelection(kCascadeDcaDaughMax, cascadeSelectionNames.at(kCascadeDcaDaughMax), config.cascadeDcaDauMax.value, limits::kUpperLimit, true, true, false);
     this->addSelection(kLambdaCpaMin, cascadeSelectionNames.at(kLambdaCpaMin), config.lambdaCpaMin.value, limits::kLowerLimit, true, true, false);
@@ -307,7 +311,11 @@ class CascadeSelection : public baseselection::BaseSelection<float, o2::analysis
   {
     this->reset();
     // cascade selections
-    this->evaluateObservable(kCascadeCpaMin, cascade.casccosPA(col.posX(), col.posY(), col.posZ()));
+    const float cpa = cascade.casccosPA(col.posX(), col.posY(), col.posZ());
+    this->updateLimits(kCascadeCpaMin, cascade.pt());
+    this->evaluateObservable(kCascadeCpaMin, cpa);
+    this->updateLimits(kCascadePaMax, cascade.pt());
+    this->evaluateObservable(kCascadePaMax, std::acos(cpa));
     this->evaluateObservable(kCascadeDcaDaughMax, cascade.dcacascdaughters());
     this->evaluateObservable(kCascadeTransRadMin, cascade.cascradius());
 

